File: 203 CS Inherited Events (Chien)/class Meute.cs
using System;
using System.Collections.ObjectModel;


namespace CS203
{
    class Meute<T> where T : Animal
    {
        private readonly Collection<T> mCol;

        public Meute()
        {
            mCol = new Collection<T>();
        }

        public Meute(T a1)
        {
            // Using a Collection Initializer
            mCol = new Collection<T> { a1 };
        }

        public void Add(T a1)
        {
            mCol.Add(a1);
        }

        public void Enerver()
        {
            foreach (T a in mCol)
            {
                a.Enerver();
            }
        }
    }


    // Custom Event Handler in C#
    class Events
    {
        event EventHandler PreDrawEvent;

        event EventHandler OnDraw
        {
            add
            {
                lock (PreDrawEvent)
                {
                    PreDrawEvent += value;
                }
            }
            remove
            {
                lock (PreDrawEvent)
                {
                    PreDrawEvent -= value;
                }
            }
        }

    }

}

File: Net9/100-209/203 VB Inherited Events (Chien)/class Meute.vb
' 2021-09-19 	PV		VS2022; Net6
' 2023-01-10	PV		Net7
' 2023-11-18	PV		Net8
' 2024-11-15	PV		Net9

Imports System.Collections.ObjectModel

Friend Class Meute(Of T As Animal)
    Private ReadOnly mCol As Collection(Of T)

    Public Sub New()
        mCol = New Collection(Of T)
    End Sub

    Public Sub New(a1 As T)
        mCol = New Collection(Of T) From {
            a1
        }
    End Sub

    Public Sub Add(a1 As T)
        mCol.Add(a1)
    End Sub

    Public Sub Enerver()
        For Each a As T In mCol
            a.Enerver()
        Next
    End Sub

End Class

'' Custom Event Handler in VB
'Class Events
'    Private Event PreDrawEvent As EventHandler

'    Private Custom Event OnDraw As EventHandler
'        AddHandler(ByVal value As EventHandler)
'            SyncLock PreDrawEvent
'                AddHandler PreDrawEvent, value
'            End SyncLock
'        End AddHandler
'        RemoveHandler(ByVal value As EventHandler)
'            SyncLock PreDrawEvent
'                RemoveHandler PreDrawEvent, value
'            End SyncLock
'        End RemoveHandler
'    End Event
'
'End Class
